merkle_tree/src/main.rs

```diff
@@ -0,0 +1,432 @@
+
+        Self {
+            tree: current_tree.clone(),
+            leaves: data_vector.len(),
+            og_leaves: string_data_vec.clone(),
+        }
+    }
+
+    /// Given a proof and an index of a data hash the function must determine wether the proof is correct.
+    /// For understanding this function we can consider the following tree:
+    ///               root
+    ///          h12        h34
+    ///        h1   h2    h3   h4
+    ///        |    |     |     |
+    ///     data1 data2 data3 data4
+    /// The array representation would be: [root,h12,h34,h1,h2,h3,h4]
+    /// With the indexes                     0   1   2   3  4  5   6
+    fn verify(&self, proofs: Vec<String>, mut leaf_index: usize) -> bool {
+        let mut hash_for_verification = self.tree[leaf_index].clone();
+        for proof in proofs {
+            if leaf_index % 2 == 1 {
+                // if index is odd we are on a left branch, so the verification must be computed concatenating the proof second
+                hash_for_verification = digest(hash_for_verification.to_owned() + &proof);
+            } else {
+                // if index is even we are on a right branch, so the verification must be computed concatenating the proof first
+                hash_for_verification = digest(proof.clone() + &hash_for_verification);
+                leaf_index -= 1; //if it's a right child this is necesary to calculate its parent
+            }
+            leaf_index = leaf_index / 2;
+        }
+        hash_for_verification == self.tree[0]
+    }
+
+    /// Given an index of a data hash the function must return the proof that the tree contains that data hash.
+    fn generate_proof(&self, mut node_index: usize) -> Option<Vec<String>> {
+        let mut proof: Vec<String> = vec![];
+        if node_index > self.tree.len() {
+            return None;
+        }
+        while node_index > 0 {
+            if node_index % 2 == 1 {
+                // if index is odd we are on a left branch, so the verification must be computed concatenating the proof second
+                proof.push(self.tree[node_index + 1].clone());
+            } else {
+                // if index is even we are on a right branch, so the verification must be computed concatenating the proof first
+                proof.push(self.tree[node_index - 1].clone());
+                node_index -= 1; //if it's a right child this is necesary to calculate its parent
+            }
+            node_index = node_index / 2;
+        }
+        Some(proof)
+    }
+
+    /// Returns a merkle tree with a new node, if the total amount of data is not a power of 2
+    /// the tree repeats the last leaf to build the base of the tree.
+    fn add_node(&mut self, data: &str) {
+        let mut new_data_vector: Vec<&str> = self.og_leaves.iter().map(String::as_str).collect();
+        new_data_vector.push(data);
+        let new_tree = Self::new(new_data_vector);
+        self.tree = new_tree.tree.clone();
```

Comment:
Cloning the tree each time we add a node looks expensive, let's double check if we can avoid it.